components/clarity-lsp/src/vscode_bridge.rs

```diff
@@ -0,0 +1,212 @@
+                } else if let Some(manifest_location) = get_manifest_location(&uri) {
+                    LspNotification::ManifestOpened(manifest_location)
+                } else {
+                    log!("Unsupported file opened");
+                    return Promise::resolve(&JsValue::null());
+                };
+
+                let editor_state = self.editor_state.clone();
+                let send_diagnostic = self.client_diagnostic_tx.clone();
+
+                return future_to_promise(async move {
+                    let mut result = match editor_state.try_write() {
+                        Ok(mut state) => {
+                            process_notification(command, &mut state, Some(&file_accessor)).await
+                        }
+                        Err(_) => return Err(JsValue::from("unable to lock editor_state")),
+                    };
+
+                    let mut aggregated_diagnostics = vec![];
+                    if let Ok(ref mut response) = result {
+                        aggregated_diagnostics.append(&mut response.aggregated_diagnostics);
+                    }
+
+                    for (location, mut diags) in aggregated_diagnostics.into_iter() {
+                        if let Ok(url) = Url::parse(&location.to_string()) {
+                            let value = PublishDiagnosticsParams {
+                                uri: url,
+                                diagnostics: clarity_diagnostics_to_lsp_type(&mut diags),
+                                version: None,
+                            };
+
+                            let value = match encode_to_js(&value) {
+                                Ok(value) => value,
+                                Err(_) => return Err(JsValue::from("unable to encode value")),
+                            };
+
+                            let _ = send_diagnostic.call1(&JsValue::null(), &value);
+                        }
+                    }
+                    Ok(JsValue::TRUE)
+                });
+            }
+
+            DidSaveTextDocument::METHOD => {
+                let params: DidSaveTextDocumentParams = match decode_from_js(params) {
+                    Ok(params) => params,
+                    _ => return Promise::resolve(&JsValue::null()),
+                };
+                let uri = &params.text_document.uri;
+                log!("> saved: {}", uri);
+
+                let command = if let Some(contract_location) = get_contract_location(uri) {
+                    LspNotification::ContractChanged(contract_location)
+                } else if let Some(manifest_location) = get_manifest_location(uri) {
+                    LspNotification::ManifestChanged(manifest_location)
+                } else {
+                    log!("Unsupported file opened");
+                    return Promise::resolve(&JsValue::null());
+                };
+
+                let editor_state = self.editor_state.clone();
```

Comment:
I was confused for a moment. `self.editor_state` is a `Arc<RwLock<EditorState>>`. Should this be reflected in the name?

Reply:
Yes good idea, would `editor_state_rw_lock` (or just `editor_state_lock`?) be a better name?